diff --git a/pkg/validate/container.go b/pkg/validate/container.go
index 1073b772..c840372a 100644
--- a/pkg/validate/container.go
+++ b/pkg/validate/container.go
@@ -513,7 +513,7 @@ func verifyExecSyncOutput(c internalapi.RuntimeService, containerID string, comm
 
 // createHostPath creates the hostPath and flagFile for volume.
 func createHostPath(podID string) (hostPath string) {
-	hostPath, err := os.MkdirTemp("", "test"+podID)
+	hostPath, err := os.MkdirTemp(os.Getenv("HOME"), "test"+podID)
 	framework.ExpectNoError(err, "failed to create TempDir %q: %v", hostPath, err)
 
 	flagFile := "testVolume.file"
